### PR TITLE
MXfp8 Inference Fixes and clean up

### PR DESCRIPTION
**Pull request by wdykas**
- [x] I, the PR author, have personally reviewed every line of this PR.
  - Align fixed MoE activation buffers with the 128-row swizzled-scale layout, fixing
  `scaled_grouped_mm` failures caused by mismatched data and scale rows.
  - Preserve and validate backend-specific MXFP8 scale formats across model loading,
  refit, and expert-weight stacking.
  - Restore raw scale bytes to the E8M0 view required by PyTorch scaled GEMMs.
  - Keep persistent MXFP8 buffers mutable and update them in place so CUDA-graph-
  captured addresses remain stable across refits.
  - Resolve the quantization format from the configured GEMM backend: `torch` uses
  the Triton producer, FlashInfer uses its native format, and unsupported vLLM MXFP8
  is rejected.
  - Round fused squared-ReLU activations to BF16 before quantization, matching
  training and unfused inference MXFP8 bins.
  - Replace duplicated MXFP8 layout values with shared constants.
# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/moe/activations.py [hunk 1]
```diff
@@ -226,7 +226,11 @@ def _squared_relu_quantize_kernel(
                 # Load and apply squared ReLU
                 x = tl.load(input_ptr + row * K + offs, mask=mask, other=0.0).to(tl.float32)
                 relu = tl.maximum(x, 0.0)
-                activated = relu * relu
+                # Match the unfused training/inference contract exactly:
+                # squared ReLU materializes a BF16 activation before the next
+                # MXFP8 quantization. Keeping the product in FP32 here changes
+                # quantization bins and compounds across MoE layers.
+                activated = (relu * relu).to(tl.bfloat16).to(tl.float32)
```
Comment (wdykas) [inline megatron/core/inference/moe/activations.py:236]: @sidsingh-nvidia do you think this make sense? I makes the log prob difference much lower

**Comment by wdykas** [comment]
/ok to test [b79e2ea](https://github.com/NVIDIA/Megatron-LM/commit/b79e2ea56d6b6867b261d8ceebaa46a790a5f973)

File: megatron/core/transformer/moe/moe_layer.py [hunk 2]
```diff
@@ -353,12 +354,11 @@ def __init__(
 
         # Inference-optimized mode setup
         if config.transformer_impl == "inference_optimized":
-            if config.inference_grouped_gemm_backend == 'auto':
+            if config.inference_grouped_gemm_backend == InferenceGroupedGemmBackend.FLASHINFER:
```
Comment (wdykas) [inline megatron/core/transformer/moe/moe_layer.py:357]: I dont think we previously checking anything here

File: megatron/inference/utils.py [hunk 3]
```diff
@@ -109,14 +112,7 @@ def get_model_for_inference() -> MegatronModule:
...
     if args.transformer_impl == "inference_optimized" and args.fp8_recipe == "mxfp8":
-        backend = args.inference_grouped_gemm_backend
-        if backend == "auto" or backend == "torch":
-            quant_backend = "triton"
-        elif backend == "te":
-            raise ValueError(
-                "MXFP8 quantization is not supported with "
-                "inference_grouped_gemm_backend='te'."
-            )
+        quant_backend = resolve_mxfp8_backend(args.inference_grouped_gemm_backend)
```
Comment (wdykas) [inline megatron/inference/utils.py:115]: made a helper just because we use in a few places

**Comment by wdykas** [comment]
/ok to test [4fbf9c2](https://github.com/NVIDIA/Megatron-LM/commit/4fbf9c21b34f150fa298d5f628425df1e55261c2)

File: megatron/core/inference/moe/activations.py (hunk 1 above)
Comment (santhnm2) [reply megatron/core/inference/moe/activations.py:236]: Can we have Ganesh look at this? I don't follow the intuition here but I guess if it helps empirically it must be doing something right

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/inference/moe/activations.py (hunk 1 above)
Comment (wdykas) [reply megatron/core/inference/moe/activations.py:236]: its just matching training side how numerics are done.  You just get different rounding but Ill ping Ganesh

File: megatron/core/inference/quantization/mxfp8_quantize.py [hunk 4]
```diff
@@ -171,18 +175,22 @@ def mxfp8_quantize(x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
-    assert K % 32 == 0, f"K ({K}) must be divisible by 32"
+    assert K % MXFP8_BLOCK_SIZE == 0, (
+        f"K ({K}) must be divisible by {MXFP8_BLOCK_SIZE}"
+    )
 
-    scale_cols = K // 32
-    n_row_blocks = _ceil_div(M, 128)
-    n_col_blocks = _ceil_div(scale_cols, 4)
-    total_scale_bytes = n_row_blocks * n_col_blocks * 512
+    scale_cols = K // MXFP8_BLOCK_SIZE
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/quantization/mxfp8_quantize.py:180]: is this a bugfix or a refactor?

**Review by sidsingh-nvidia — APPROVED** [review]

File: megatron/core/inference/quantization/mxfp8_quantize.py (hunk 4 above)
Comment (wdykas) [reply megatron/core/inference/quantization/mxfp8_quantize.py:180]: refactor to try to bring clarity since I was a bit confused at the time but there is a bug related to rounding. The rounding bug was the MXFP8 MoE activation-buffer row mismatch. The swizzled scale layout always pads rows to 128-row blocks. For Nano-v3, scales had 8,704 rows, while the FP8 data buffer had only 8,624 rows. scaled_grouped_mm broke on the shapes. We fixed it by rounding the MXFP8 permuted output capacity up to 128 rows.

**Review by shanmugamr1992 — APPROVED** [review]

**Review by yaox12 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [f97636b](https://github.com/NVIDIA/Megatron-LM/commit/f97636bf9b014d2dfba40e52ba74cb9262a3e419)

**Comment by wdykas** [comment]
/ok to test [80ae99b](https://github.com/NVIDIA/Megatron-LM/commit/80ae99befe80400004a651a5a40ec7e81332a06c)

**Comment by wdykas** [comment]
/ok to test [e4c8d88](https://github.com/NVIDIA/Megatron-LM/pull/6588/commits/e4c8d888406dc3053a92ed59526559d2ccd273bd)
